### PR TITLE
feat!: Allow "Row Variables" declared as List<Type>

### LINKED ISSUE #787: Allow TypeArg for a row of types
TypeParam::List / TypeArg::List allows one to declare a list of types, and provide them as type args (also other kinds of list, not relevant here). But there's no way to use these outside a custom binary function. We should allow should a parameter to be used inside a tuple, sum, or function type (inputs or outputs), or anywhere a row is expected, so one can say e.g. `unpack_tuple` has a `TypeParam::List(TypeParam::Type)` and then inputs `Tuple(TypeArg::new_variable(0, TypeParam::List(TypeParam::Type))` and outputs that type variable.

### PR DESCRIPTION
**Pull request by acl-cqc**
* Add a RowVariable variant of Type(Enum) that stands for potentially multiple types, created via Type::new_row_var_use.
* This can appear in a TypeRow, i.e. the TypeRow is then of variable length; and can be instantiated with a list of types (including row vars) or a single row variable
* Validation enforces that RowVariables are not used directly as wire/port types, but can appear *inside* other types
* OpDef's may be polymorphic over RowVariables (allowing "varargs"-like operators equivalent to e.g. MakeTuple); these must be replaced by non-rowvar types when instantiating the OpDef to an OpType
* FuncDefn's/FuncDecl's may also be polymorphic over RowVariables as long as these are not directly argument/result types
* Also add TypeParam::new_sequence

closes #787

BREAKING CHANGE: Type::validate takes extra bool (allow_rowvars); renamed {FunctionType, PolyFuncType}::(validate=>validate_var_len).

### CODE REVIEW

**Comment by acl-cqc** [comment]
> The "real" way to do this would be to change TypeRow to store elements of `Either<RowVariable, Type>` (where `struct RowVariable(usize, TypeBound)`), but I think that's just gonna be too messy with type-rows everywhere.

Thinking about this, it might not be too bad (i.e. to make `TypeRow` contain a list of `RowVariableOrType`) - given enough `impl` and `Into` magic. After all, most uses are all Types, and `type_row!` is *a macro* that can jump around typechecking issues...is it worth a try, do you think?

File: src/types.rs [hunk 1]
```diff
@@ -421,6 +479,17 @@ pub(crate) mod test {
...
+        let mut ext = Extension::new(MY_EXT);
+        ext.add_type(
+            "my_custom".into(),
+            vec![],
+            "".into(),
+            TypeBound::Copyable.into(),
+        )
+        .unwrap();
+        let reg = ExtensionRegistry::try_new([PRELUDE.to_owned(), ext]).unwrap();
+        t.validate(&reg, &[]).unwrap()
```
Comment (ss2165) [inline src/types.rs:492]: is this related?

**Review by ss2165 — COMMENTED** [review]
I agree that the "proper way" is worth a quick try (time bounded!). The macro does help us but the annoying part will likely be that most uses of rows treat it like a slice of types (iterating, getting etc.). I'm imagining moving the current `TypeRow` impls to a new `ConcreteTypeRow` struct and then adding a panicking `as_concrete()` to a TypeRow enum. Maybe there is something less onerous?

**Comment by aborgna-q** [comment]
@acl-cqc What is the status of this? Did you get to try the `TypeRow` / `ConcreteTypeRow` alternative?

**Comment by acl-cqc** [comment]
> @acl-cqc What is the status of this? Did you get to try the `TypeRow` / `ConcreteTypeRow` alternative?

Yes, or at least, partly, in branch `feat/row_typevar2`. `ConcreteTypeRow` is called `TypeRow` and non-concrete is `TypeRowV`. From what I remember we didn't think this looked very promising, but perhaps with more tidying up?? Both are quite out of date, certainly I can update *this* branch but I think I'd not update `feat/row_typevar2` unless we really thought that was the best way to go...

**Comment by acl-cqc** [comment]
The value of the feature is to allow us to make operations like UnpackTuple, MakeTuple, etc., as *library* operations in the prelude, rather than the current LeafOp enum of ops which (without this PR) cannot be expressed in the extension system...

**Comment by acl-cqc** [comment]
@doug - moved discussion to #787

File: hugr/src/types.rs [hunk 2]
```diff
@@ -317,10 +313,21 @@ impl Type {
...
     /// variable was declared (i.e. as a [TypeParam::Type]`(bound)`).
-    pub fn new_var_use(idx: usize, bound: TypeBound) -> Self {
+    pub const fn new_var_use(idx: usize, bound: TypeBound) -> Self {
         Self(TypeEnum::Variable(idx, bound), bound)
     }
 
+    /// New use (occurrence) of the row variable with specified index.
+    /// `bound` must match that with which the variable was declared
+    /// (i.e. as a [TypeParam::List]` of a `[TypeParam::Type]` of that bound).
+    /// For use in [OpDef], not [FuncDefn], type schemes only.
```
Comment (acl-cqc) [inline hugr/src/types.rs:323]: This is the critical distinction, which is neither explained nor probably made, as clear(ly) as it needs to be. (So far, in this PR!)

File: hugr/src/types.rs [hunk 3]
```diff
@@ -404,20 +438,32 @@ impl<'a> Substitution<'a> {
         arg.clone()
     }
 
+    fn apply_rowvar(&self, idx: usize, bound: TypeBound) -> Vec<Type> {
+        let arg = self
+            .0
+            .get(idx)
+            .expect("Undeclared type variable - call validate() ?");
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types.rs:445]: Gone the other way (consistent with the existing one above)

File: hugr/src/types.rs [hunk 4]
```diff
@@ -404,20 +438,32 @@ impl<'a> Substitution<'a> {
...
+                .iter()
+                .map(|ta| match ta {
+                    TypeArg::Type { ty } => ty.clone(),
+                    _ => panic!("Not a list of types - did validate() ?"),
+                })
+                .collect(),
+            TypeArg::Type { ty } => {
+                debug_assert_eq!(check_type_arg(arg, &TypeParam::Type { b: bound }), Ok(()));
+                vec![ty.clone()]
+            }
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types.rs:457]: Yes, I've added a comment that (according to `new_row_var_use`) row vars should be represented as `TypeArg::Type` of a Row Variable "`Type`". And, erm, cleaned/fixed an assert that was....more permissive than it should have been

File: hugr/src/types/type_param.rs [hunk 5]
```diff
@@ -217,6 +218,17 @@ impl TypeArg {
             TypeParam::Type { b } => TypeArg::Type {
                 ty: Type::new_var_use(idx, b),
             },
+            TypeParam::List { param: bx } if matches!(*bx, TypeParam::Type { .. }) => {
+                // There are two reasonable schemes for representing row variables:
+                // 1. TypeArg::Variable(idx, TypeParam::List(TypeParam::Type(typebound)))
+                // 2. TypeArg::Type(Type::new_row_var(idx, typebound))
+                // Here we prefer the latter for canonicalization, although we cannot really
+                // prevent both if users construct the TypeArg variants directly (doing so will break Eq)
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:234]: You mean so that `TypeArg::Variable(idx, TypeParam::List(TypeParam::Type(bound)))` is `==` to `TypeArg::Type(Type::RowVariable(idx, bound))` ?

Yes, we could, that is fair I guess. I would really love to stick to a canonical representation, but that is hard without going through the type system (as per branch `feat/row_typevar2` where we had `enum RowVarOrType`, or `feat/row_typevar3` where TypeArg could contain a `Type<false>` thus ruling out the RHS of the `==` example above)

I guess we could also go for the LHS of the `==` by making `impl From<Type> for TypeArg` more complex.

File: hugr/src/types/type_param.rs [hunk 6]
```diff
@@ -217,6 +218,17 @@ impl TypeArg {
...
+                let TypeParam::Type { b } = *bx else { panic!() };
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:235]: No! Sadly both `if let` and "box patterns" (the problem is caused by being unable to match the thing inside the box inside the TypeParam::List) are "experimental" at this point :-(

[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:235]: @aborgna-q

File: hugr/src/types/type_param.rs [hunk 7]
```diff
@@ -315,6 +340,23 @@ impl CustomTypeArg {
...
+}
+
+fn check_type_arg_rv(
+    arg: &TypeArg,
+    param: &TypeParam,
+    allow_rowvars: bool,
+) -> Result<(), TypeArgError> {
+    // allow_row_vars only applies if we are checking against values allowed inside a list
+    // (where the row variable could stand for *several* elements)
+    debug_assert!(!allow_rowvars || matches!(param, TypeParam::Type { .. }));
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:359]: Yeah, realize this was quite hard to follow. I've managed to remove `allow_rowvars` and go back to just one recursive function, hope this is better.

File: hugr/src/types/type_param.rs [hunk 8]
```diff
@@ -323,13 +365,21 @@ pub fn check_type_arg(arg: &TypeArg, param: &TypeParam) -> Result<(), TypeArgErr
...
+            let allow_rvs = matches!(&**param, TypeParam::Type { .. });
+            elems
+                .iter()
+                .try_for_each(|arg| check_type_arg_rv(arg, param, allow_rvs))
         }
+        // Also allow a single "Type" to be used for a List *only* if the Type is a row variable
+        // (i.e., it's not really a Type, it's multiple Types)
+        (TypeArg::Type { ty }, TypeParam::List { param }) if rowvar_in_list(ty, param) => Ok(()),
+
         (TypeArg::Sequence { elems: items }, TypeParam::Tuple { params: types }) => {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:381]: We do not allow a *value* containing a row variable to be used in a tuple. Tuple has fixed arity, so the "row variable" would have to have exactly the correct arity to fit. In which case, it doesn't need to be a row variable.

Note that a valid `TypeArg::Type` is a `Type::Tuple(some_row_var)`, but this is a single type, and would fit only into a `TypeParam::Type`, as long as TypeBounds fit but regardless of the row-var.

File: hugr/src/types/type_row.rs [hunk 9]
```diff
@@ -78,6 +83,28 @@ impl TypeRow {
...
+    pub(super) fn substitute(&self, tr: &Substitution) -> TypeRow {
+        let res = self
+            .iter()
+            .flat_map(|ty| ty.substitute(tr))
+            .collect::<Vec<_>>()
+            .into();
+        res
+    }
+
+    pub(super) fn validate_var_len(
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_row.rs:100]: I am dead against this, sorry! Where there is ambiguity as to whether `validate` might accept row vars or not, I think it's critical that we do not use that name. Without the type system helping (by parameterization `<Type>` vs `<RowVarOrType>`, or `<true>` vs `<false>`), naming is the only thing to help us avoid mistakes. Here we *could* just as well define `validate_fixed_len`, indeed at the same time, but it just happens that the latter isn't used (an accident of the way `validate_port_kind` is structured, really)

File: hugr/src/types/signature.rs [hunk 10]
```diff
@@ -32,22 +32,21 @@ impl FunctionType {
         self
     }
 
-    pub(crate) fn validate(
+    pub(super) fn validate_varargs(
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/signature.rs:35]: This one can be handled by the Signature/FunctionType distinction, coming up

File: hugr/src/types/poly_func.rs [hunk 11]
```diff
@@ -73,9 +73,12 @@ impl PolyFuncType {
 
     /// Validates this instance, checking that the types in the body are
     /// wellformed with respect to the registry, and the type variables declared.
-    pub fn validate(&self, reg: &ExtensionRegistry) -> Result<(), SignatureError> {
+    /// Allows both inputs and outputs to contain [RowVariable]s
+    ///
+    /// [RowVariable]: [crate::types::TypeEnum::RowVariable]
+    pub fn validate_varargs(&self, reg: &ExtensionRegistry) -> Result<(), SignatureError> {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/poly_func.rs:79]: This is a stranger case than [TypeRow](https://github.com/CQCL/hugr/pull/804/files/7214cbc328f0991e5babc0d9be2144d9050619b4#r1599919712). We *do* have a use for a PolyFuncType that doesn't allow row vars: the signature of a FuncDecl/FuncDefn (i.e. inside the binders). However, it happens that we never validate that, and I'm much less sure why we don't. Hmmm.

If we can distinguish Signature from FunctionType, then we might be able to distinguish the two variants of PolyFuncType, too. Lemme see....

File: hugr/src/types/type_param.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:233]: Hmmmm, so I was thinking that since `TypeArg` is `pub enum`, they could construct anything. *However* `TypeArg::Variable` contains a `pub struct TypeArgVariable` and in fact this *cannot* be publically constructed - it has private members and no `new`.

Hence, the only way to construct a `TypeArg::Variable` from outside the `types` module is by `new_var_use` (or `new_row_var_use`), which means we *can* control canonicalization. Which is to say, I think the comment here is bogus.

Do you agree?

However, there is still the possibility of canonicalizing in the other direction (no `Type::RowVariable` inside `TypeArg::Type`, use `TypeArgVariable` instead). I guess I could experiment with a switcheroo commit....

File: hugr/src/types/type_param.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:234]: We would need something like `TypeParam::List { param: box TypeParam::Type { b } } => ...` and the `box` there is not allowed yet

File: hugr/src/types/type_param.rs [hunk 12]
```diff
@@ -323,13 +365,21 @@ pub fn check_type_arg(arg: &TypeArg, param: &TypeParam) -> Result<(), TypeArgErr
             _,
         ) if param.contains(cached_decl) => Ok(()),
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:354]: Not by current canonicalization as `TypeArg::Type(Type::RowVariable)` (which is to say, `cached_decl` cannot be a `TypeParam::List { TypeParam::Type }` - and I wasn't planning on doing row-variables that aren't types, the use cases are tiny. If we canonicalized in the other direction, hmmm, maybe that would fall out, hmmm)

File: hugr/src/types/poly_func.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/poly_func.rs:79]: Yah you could never link in a single FuncDefn, but you could replace every Call to it with a different Call (with some typeargs missing, and targetting a different FuncDefn with those typeargs "fixed").

Is that really useful? I feel a bit unconvinced. Maybe one should just have different FuncDecls to start with?

File: hugr/src/types/type_row.rs [hunk 13]
```diff
@@ -40,6 +43,8 @@ impl TypeRow {
 
     #[inline(always)]
     /// Returns the port type given an offset. Returns `None` if the offset is out of bounds.
+    // Note/TODO: it might be good to disable this if we are indexing over (past) a Row Variable,
+    // as substitution could change where in the row the offset refers.
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_row.rs:47]: I started in branch `feat/row_typevar4` (!). I remember from last time that it was *easier* to change the variable-arity FunctionType (to FunTypeVarArgs) that the fixed-arity one (to Signature), even if the ideal is probably the other way around, so that's what I started doing. But it's still very hard: you need to change every `FunctionType::new` into either `FunTypeVarArgs::new` or `FunctionType::try_new....unwrap` (as FunctionType::new doesn't exist, it needs to check against row vars).

Maybe I've got it wrong and changing fixed-arity to Signature(::try_new) is the "easier" way, I dunno. I am minded to leave this and think about coming back to it as a separate/second PR ?

File: hugr/src/types/type_param.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:376]: I have added some tests...not convinced that rstest `cases` help very much, might do better without

[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:376]: And some more, abandoning `rstest`

File: hugr/src/types/type_param.rs [hunk 14]
```diff
@@ -315,6 +340,23 @@ impl CustomTypeArg {
 
 /// Checks a [TypeArg] is as expected for a [TypeParam]
 pub fn check_type_arg(arg: &TypeArg, param: &TypeParam) -> Result<(), TypeArgError> {
+    check_type_arg_rv(arg, param, false)
+}
+
+fn check_type_arg_rv(
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs]: Added a load of tests.

Those in `poly_func.rs` `test_bound_covariance` could probably move here too.

File: hugr/src/types.rs [hunk 15]
```diff
@@ -344,47 +351,74 @@ impl Type {
...
+    /// Checks that this [Type] represents a single Type, not a row variable,
+    /// that all variables used within are in the provided list of bound variables,
+    /// and that for each [CustomType], the corresponding
     /// [TypeDef] is in the [ExtensionRegistry] and the type arguments
     /// [validate] and fit into the def's declared parameters.
     ///
     /// [validate]: crate::types::type_param::TypeArg::validate
     /// [TypeDef]: crate::extension::TypeDef
-    pub(crate) fn validate(
+    pub(crate) fn validate_1type(
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types.rs:370]: Hmmm, I also considered `validate_not_rowvar` but decided that was longer...I'm very happy to consider any alternative name that is clear whether it accepts rowvars or not ;-)

File: hugr/src/types/poly_func.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/poly_func.rs:79]: But the handling of FuncDecl's is only really relevant if we want to go with branch `feat/row_typevar4` (splitting FunctionType and PolyFuncType), which I don't think we should do here. (As a second step if we want to at all.)

I will stick by not using `validate` - even if `validate_fixed_len` is not defined here, it would be equally valid (just that we don't use it). Happy to entertain any non-ambiguous alternative to `validate_varargs` tho.

File: hugr/src/types/signature.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/signature.rs:35]: No, deferred to branch `feat/row_typevar4`, prefer not `validate`as elsewhere

**Comment by acl-cqc** [comment]
> Can you add some tests to `hugr::serialize::test`?

Added a couple more cases to `roundtrip_polyfunctype`

File: hugr/src/types/type_row.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_row.rs:47]: See #1044 (not building yet)

File: hugr/src/types.rs [hunk 16]
```diff
@@ -404,20 +446,34 @@ impl<'a> Substitution<'a> {
...
     }
 
+    fn apply_rowvar(&self, idx: usize, bound: TypeBound) -> Vec<Type> {
+        let arg = self
+            .0
+            .get(idx)
+            .expect("Undeclared type variable - call validate() ?");
+        debug_assert_eq!(check_type_arg(arg, &TypeParam::new_list(bound)), Ok(()));
+        match arg {
+            // Row variables are represented as 'TypeArg::Type's (see TypeArg::new_row_var_use)
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types.rs:456]: Yes. It'd be great if there was some way to get `cargo doc` to check these links, but starting with `///` is not sufficient. Thanks!

File: hugr/src/types/type_param.rs [hunk 17]
```diff
@@ -214,9 +223,16 @@ impl TypeArg {
     /// `bound` must match that with which the variable was declared.
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:223]: Even there it must have been declared *by* the polymorphic function (a FuncDefn or FuncDecl has a list of TypeParam binders)...

However, I'll change "must match" to "must be exactly" to be clear

File: hugr/src/types/type_param.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:363]: However ATM I am leaning against canonicalizing `TypeArg::Type(Type::RowVariable)` to `TypeArg::Variable(...TypeParam::List(TypeParam::Type))`, or indeed `TypeArg::Type(Type::Variable)` to `TypeArg::Variable(TypeParam::Type)` because (as well as then needing a TypeNotVariable to go inside `TypeArg::Type`) we would *still* need to keep `TypeArgVariable` to rule out `TypeArg::Variable(TypeParam::Extensions)` which should be an `ExtensionSet::new_type_var`

File: hugr/src/types.rs [hunk 18]
```diff
@@ -344,47 +351,82 @@ impl Type {
...
-                rows.iter().flat_map(|x| x.iter()),
-            ),
+            TypeEnum::Sum(SumType::General { rows }) => rows
+                .iter()
+                .try_for_each(|row| row.validate_var_len(extension_registry, var_decls)),
             TypeEnum::Sum(SumType::Unit { .. }) => Ok(()), // No leaves there
             TypeEnum::Alias(_) => Ok(()),
             TypeEnum::Extension(custy) => custy.validate(extension_registry, var_decls),
-            TypeEnum::Function(ft) => ft.validate(extension_registry, var_decls),
+            // FunctionTypes used as types of values may have unknown arity (e.g. if they are not called)
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types.rs:384]: I think "Function *values*" is what I was looking for, without needing to mention `FunctionType`

[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types.rs:384]: I think "Function *values*" is what I was looking for, without needing to mention `FunctionType`

File: hugr/src/types/type_param.rs [hunk 19]
```diff
@@ -258,9 +275,17 @@ impl TypeArg {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:289]: Ok, I agree checking in validation is a good move. There are three options:
* A new SignatureError variant. I.e. a "public" error. This is the "nicest", but if this occurs, this is a coding bug on our part rather than a user / Hugr construction error, right? I mean, even deserialization should use `new_var_use` etc. rather than directly constructing `TypeArgVariable`. Hence....
* An `assert`. This is what I've done, but it may be the worst of three options; if this occurs in "production", user will not only use Eq-totalness, but indeed, the rest of their program run too ;-).
* `debug_assert`

I went for the second but I'm moving towards the third, unless you strongly push me towards the first...

[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/type_param.rs:289]: Re `check_typevar_decl`, we can't do it there as the *correct* representation as a `TypeArg::Type` of a `Type(Enum)::(Row)Variable` eventually comes back to `check_typevar_decl` with the correct/expected form of TypeParam(::List). So I think that's in the right place - variables bound by TypeParams can be just Types not TypeArgs.

File: hugr/src/types.rs [hunk 20]
```diff
@@ -326,6 +322,17 @@ impl Type {
         Self(TypeEnum::Variable(idx, bound), bound)
     }
 
+    /// New use (occurrence) of the row variable with specified index.
+    /// `bound` must match that with which the variable was declared
+    /// (i.e. as a [TypeParam::List]` of a `[TypeParam::Type]` of that bound).
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types.rs]: Changed to "must be exactly". The arg and param may be `Eq` when the arg is (put into a hole requiring an `Any`) aka (used as an `Any`), but arg must be the same as param.

File: hugr/src/types.rs [hunk 21]
```diff
@@ -344,47 +351,82 @@ impl Type {
         TypeBound::Copyable.contains(self.least_upper_bound())
     }
 
-    /// Checks all variables used in the type are in the provided list
-    /// of bound variables, and that for each [CustomType] the corresponding
+    /// If this Type is a row variable, return its bound, otherwise None
+    pub fn row_var_bound(&self) -> Option<TypeBound> {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types.rs]: Indeed, not quite sure what I was thinking, I guess it seemed like a good idea at the time ;-). Done, thanks

**Comment by acl-cqc** [comment]
> I am scared by the FunctionType::port_type methods, I think we should remove mention of ports from TypeRow and add debug_assert! to FunctionType::port_type.

Yes, I am too. I have added `debug_assert` to the TypeRow methods (into which everything else eventually calls).

I did consider making the TypeRow methods non-`pub`, in b9fc1cf92452a0ae7e33a0f3159704c9ccdf0abe, but since they have the (debug) checks in them they are no worse than any others, and any removal affects the public API (and otherwise there are no breaking changes here, IIUC). So I completely agree these are nasty and to be avoided but I think that is the brief for #1044 - which *is* a breaking change (but only adds protection, not functionality)....

File: hugr/src/types/type_param.rs [hunk 22]
```diff
@@ -252,15 +270,36 @@ impl TypeArg {
...
     }
 
     pub(crate) fn substitute(&self, t: &Substitution) -> Self {
         match self {
-            TypeArg::Type { ty } => TypeArg::Type {
-                ty: ty.substitute(t),
-            },
+            TypeArg::Type { ty } => {
+                // A row variable standing for many types is represented as a single type
+                // TODO: this case can't happen until we start substituting across Hugrs
```
Comment (acl-cqc) [inline hugr/src/types/type_param.rs:294]: I'm not sure this "can't happen until". But I haven't come up with a case yet. While *TypeDef*s do not allow row vars (only *OpDef*s), *and* we haven't tackled substitution across nodes (#709) it may indeed be that it can't happen. But, I've captured it in a unit test, and the impl here was broken, so fixed.

**Comment by acl-cqc** [comment]
Also (note to self): should add a test of `TypeParam::List(TypeParam::List(TypeParam::Type))` to check I haven't messed that up accidentally.

File: hugr/src/types/type_param.rs [hunk 23]
```diff
@@ -396,3 +469,172 @@ pub enum TypeArgError {
...
+    #[test]
+    fn type_arg_subst_row() {
+        let row_param = TypeParam::new_list(TypeBound::Copyable);
+        let row_arg = TypeArg::Sequence {
+            elems: vec![BOOL_T.into(), Type::UNIT.into()],
+        };
+        check_type_arg(&row_arg, &row_param).unwrap();
+
+        // Now say a row variable referring to *that* row was used
+        // to instantiate an outer "row parameter" (list of type).
```
Comment (acl-cqc) [inline hugr/src/types/type_param.rs:560]: So I think I can do a non-unit-test of this - it requires defining a custom `TypeDef` parameterized by a list-of-type, then a Hugr FuncDef (parameterized by a list-of-type) returning an instance of that type-def (using the function's row-variable parameter). Would be significantly longer but I think it would work. Worth doing???

**Comment by acl-cqc** [comment]
* Removed "Port" from TypeRow on main in #1087, now merged in here
* Moved not-indexing-over-rowvars check from TypeRow to FunctionType via helper
* Type::validate{_1type,_in_row} renamed/combined to validate with extra bool first argument
* FunctionType/PolyFuncType `validate_varargs` renamed to `validate_var_len` for consistency with TypeRow

**Comment by acl-cqc** [comment]
Oh, and fixed a failing proptest using a new `Type::non_row_var` strategy for the element type of list *values*. This is sufficient for the test, but still (generally speaking) far from valid; a list-value must have a closed type without any type variables at all (row or type vars, at the top level or elsewhere) whereas this only rules out top-level row vars.

File: hugr/src/types/signature.rs [hunk 24]
```diff
@@ -193,6 +193,17 @@ impl FunctionType {
 }
 
 impl FunctionType {
+    /// If this FunctionType contains any row variables, return one.
+    pub fn find_rowvar(&self) -> Option<(usize, TypeBound)> {
```
[the remark this answers is not in the document]
Comment (acl-cqc) [reply hugr/src/types/signature.rs:197]: I was going to have `has_rowvars(&self) -> bool` and thought that definitely *should* be `pub`. Then I realized that when it returned `true` I'd have to scan through to find one to raise the error I wanted, so why not do that in the same go. I think having some kind of are-there-any-rowvars is a good method to have, I agree this is a slightly non-obvious form for it but it's no worse than the `bool` one no?

Could return a list of vars, or `Vec<(Direction, usize, (usize, TypeBound)>` (i.e. *where* each one was found) or something like that I guess?
